server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,378 @@
+                "it requires executing in [depth_first] mode, which the VariableWidthHistogram agg cannot do.");
+        }
+    }
+
+    private String subAggsNeedScore() {
+        for (Aggregator subAgg : subAggregators) {
+            if (subAgg.scoreMode().needsScores()) {
+                return subAgg.name();
+            }
+        }
+        return null;
+    }
+
+    private String descendsFromNestedAggregator(Aggregator parent) {
+        while (parent != null) {
+            if (parent.getClass() == NestedAggregator.class) {
+                return parent.name();
+            }
+            parent = parent.parent();
+        }
+        return null;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        if (valuesSource != null && valuesSource.needsScores()) {
+            return ScoreMode.COMPLETE;
+        }
+        return super.scoreMode();
+    }
+
+    @Override
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return true;
+    }
+
+    @Override
+    public DeferringBucketCollector getDeferringCollector() {
+        deferringCollector = new MergingBucketsDeferringCollector(context);
+        return deferringCollector;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
+        return new LeafBucketCollectorBase(sub, values){
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                assert bucket == 0;
+                if(values.advanceExact(doc)){
+                    final int valuesCount = values.docValueCount();
+                    double prevVal = Double.NEGATIVE_INFINITY;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        double val = values.nextValue();
+                        assert val >= prevVal;
+                        if (val == prevVal){
+                            continue;
+                        }
+                        collectValue(sub, doc, val);
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * A new bucket is created for the first <code>cacheLimit</code> documents.
+     * Then, these documents are clustered into <code>3/4 * shardSize</code> buckets.
+     * Subsequent documents will either be be collected into the nearest existing bucket, or into a new bucket if
+     * their value is distant from all existing buckets.
+     */
+    private void collectValue(LeafBucketCollector sub, int doc, double val) throws IOException{
+        if (numCachedDocs < cacheLimit) {
+            // Make a new bucket for the document
+            cachedValues = bigArrays.resize(cachedValues, cachedValues.size() + 1);
```

Comment:
This is going to resize the underlying storage every time a new bucket comes in. `grow` will oversize the array which will amortize the allocations better. It'll oversize beyond what you need, but that seems like a price that'd be ok to pay.